src/algorithms/calorimetry/TrackClusterMergeSplitter.cc

```diff
@@ -0,0 +1,460 @@
+
+  }  // end 'process(Input&, Output&)'
+
+
+
+  // --------------------------------------------------------------------------
+  //! Reset bookkeeping containers
+  // --------------------------------------------------------------------------
+  void TrackClusterMergeSplitter::reset_bookkeepers() const {
+
+    m_mapIsConsumed.clear();
+    m_mapClustProject.clear();
+    m_mapClustToMerge.clear();
+    m_mapProjToMerge.clear();
+    m_vecProject.clear();
+    m_vecClust.clear();
+    trace("Reset bookkeeping containers");
+
+  }  // end 'reset_bookkeepers()'
+
+
+
+  // --------------------------------------------------------------------------
+  //! Collect projections pointing to calorimeter
+  // --------------------------------------------------------------------------
+  void TrackClusterMergeSplitter::get_projections(
+    const edm4eic::TrackSegmentCollection* projections,
+    const edm4eic::CalorimeterHit& hit
+  ) const {
+
+    // return if projections are empty
+    if (projections -> size() == 0) {
+      debug("No projections in input collection.");
+      return;
+    }
+
+    // get readout
+    auto readout = m_converter -> findReadout(
+      m_converter -> findDetElement(
+        m_converter -> position(hit.getCellID())
+      )
+    );
+
+    // grab detector id
+    const int id = m_detector -> constant<int>(
+      std::regex_replace(readout.name(), std::regex("Hits"), "_ID")
+    );
+    debug("Collecting projections to detector with system id {}", id);
+
+    // collect projections
+    for (auto project : *projections) {
+      for (auto point : project.getPoints()) {
+        if (
+          (point.system  == id) &&
+          (point.surface == 1)
+        ) {
+          m_vecProject.push_back(point);
+        }
+      }  // end point loop
+    }  // end projection loop
+    trace("Collected relevant projections: {} to process", m_vecProject.size());
+
+  }  // end 'get_projections(edm4eic::CalorimeterHit&, edm4eic::TrackSegmentCollection&)'
+
+
+
+  // --------------------------------------------------------------------------
+  //! Match clusters to track projections
+  // --------------------------------------------------------------------------
+  /*! FIXME this might be better handled in a separate algorithm
+   */
+  void TrackClusterMergeSplitter::match_clusters_to_tracks(
+    const edm4eic::ProtoClusterCollection* clusters
+  ) const {
+
+
+    // loop over relevant projections
+    for (uint32_t iProject = 0; iProject < m_vecProject.size(); ++iProject) {
+
+      // grab projection
+      auto project = m_vecProject[iProject];
+
+      // get eta, phi of projection
+      const float etaProj = edm4hep::utils::eta(project.position);
+      const float phiProj = atan2(project.position.y, project.position.x);
+
+      // find closest cluster
+      bool  foundMatch = false;
+      float dMatch = std::numeric_limits<float>::max();
+      int   iMatch = std::numeric_limits<int>::max();
```

Comment:
```suggestion
      float dMatch;
      int   iMatch;
```
foundMatch should be enough to prevent UB in the current code. If somebody ever breaks, UB is preferable to inserting nonsense, to my taste.